parsers/NG.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python3
+
+"""Parser for the electricity grid of Nigeria"""
+
+import arrow
+import logging
+import requests
+from bs4 import BeautifulSoup
+
+
+LIVE_PRODUCTION_API_URL = "https://www.niggrid.org/GenerationLoadProfileBinary?readingDate={0}&readingTime={1}"
+
+TYPE_MAPPING = {"hydro": "hydro", "gas": "gas", "gas/steam": "gas", "steam": "gas"}
+
+
+def extract_name_tech(company):
+    parts = company.split("(")
+    tech = parts[1].strip(")").lower()
+
+    return parts[0], TYPE_MAPPING[tech]
+
+
+def template_response(zone_key, datetime, source):
+    return {
+        "zoneKey": zone_key,
+        "datetime": datetime.isoformat(),
+        "production": {
+            "biomass": 0.0,
+            "coal": 0.0,
+            "gas": 0.0,
+            "hydro": 0.0,
+            "nuclear": 0.0,
+            "oil": 0.0,
+            "solar": 0.0,
+            "wind": 0.0,
+            "geothermal": 0.0,
+            "unknown": 0.0,
+        },
+        "storage": {},
+        "source": source,
+    }
+
+
+def fetch_production(
+    zone_key=None,
+    session=None,
+    target_datetime=None,
+    logger=logging.getLogger(__name__),
+):
+    """Requests the last known production mix (in MW) of a given zone
+    Arguments:
+    zone_key (optional) -- used in case a parser is able to fetch multiple zones
+    session (optional) -- request session passed in order to re-use an existing session
+    target_datetime (optional) -- used if parser can fetch data for a specific day, a string in the form YYYYMMDD
+    logger (optional) -- handles logging when parser is run
+    Return:
+    A list of dictionaries in the form:
+    {
+      'zoneKey': 'FR',
+      'datetime': '2017-01-01T00:00:00Z',
+      'production': {
+          'biomass': 0.0,
+          'coal': 0.0,
+          'gas': 0.0,
+          'hydro': 0.0,
+          'nuclear': null,
+          'oil': 0.0,
+          'solar': 0.0,
+          'wind': 0.0,
+          'geothermal': 0.0,
+          'unknown': 0.0
+      },
+      'storage': {
+          'hydro': -10.0,
+      },
+      'source': 'mysource.com'
+    }
+    """
+
+    if target_datetime is not None:
+        raise NotImplementedError('This parser is not yet able to parse past dates')
+
+    now = arrow.now(tz="Africa/Lagos")
+    dt_day = now.format("DD/MM/YYYY")
+    dt_hm = now.format("HH:00")
+    fullUrl = LIVE_PRODUCTION_API_URL.format(dt_day, dt_hm)
+
+    r = session or requests.session()
+    resp = r.get(fullUrl)
+
+    try:
+        soup = BeautifulSoup(resp.text, "html.parser")
+        table = soup.find("table", {"id": "MainContent_gvGencoLoadProfiles"})
+        rows = table.find_all("tr")[1:-1]  # ignore header and footer rows
+    except AttributeError:
+        raise LookupError('No data currently available for Nigeria.')
+
+    result = template_response(zone_key, now, "niggrid.org")
```

Comment:
...otherwise, the `datetime` on the result is going to be different than that of the website. (E.g. fetching at 12:34 retrieves data for 12:00, but gets logged as data for 12:34 to the ElectricityMap database.)

Reply:
I've updated this such that the `datetime` returned is the current timestamp `now = arrow.now(tz="Africa/Lagos")` - this seems to be the pattern in the others parsers

Reply:
What I meant is that in this function call  `now` should be (a variable with the the result) of  `now.replace(minute=0)` so the timestamp returned by the parser is the same one as the timestamp given on the website

Reply:
I'd replace line 83-98 with
```


    timestamp = arrow.now(tz="Africa/Lagos").replace(minute=0)
    dt_day = timestamp.format("DD/MM/YYYY")
    dt_hm = timestamp .format("HH:mm")
    fullUrl = LIVE_PRODUCTION_API_URL.format(dt_day, dt_hm)

    r = session or requests.session()
    resp = r.get(fullUrl)

    try:
        soup = BeautifulSoup(resp.text, "html.parser")
        table = soup.find("table", {"id": "MainContent_gvGencoLoadProfiles"})
        rows = table.find_all("tr")[1:-1]  # ignore header and footer rows
    except AttributeError:
        raise LookupError("No data currently available for Nigeria.")

    result = template_response(zone_key, timestamp, "niggrid.org")
```

Line 83, 84, 85 and 98 are changed. (Variable name changed only for the sake of clarity)

Reply:
Ohhh.... I understand now. I've updated the PR to reflect your comment